src/cards/polymorph.ts

```diff
@@ -0,0 +1,165 @@
+              const newUnit = polymorphUnit(unit, chosenUnitType.id, underworld, prediction);
+              if (newUnit) {
+                // Visual aid only
+                if (prediction) {
+                  // TODO - Show new unit?
+                }
+
+                // Keep Modifiers
+                for (const modifierKey of Object.keys(unit.modifiers)) {
+                  const modifier = allModifiers[modifierKey];
+                  const modifierInstance = unit.modifiers[modifierKey];
+                  if (modifier && modifierInstance) {
+                    if (modifier?.add) {
+                      modifier.add(newUnit, underworld, prediction, modifierInstance.quantity, modifierInstance);
+                    }
+                  } else {
+                    console.error("Modifier doesn't exist? This shouldn't happen.");
+                  }
+                }
+
+                // Cleanup old unit and remove it from targets
+                Unit.cleanup(unit, false);
+                state.targetedUnits = state.targetedUnits.filter(u => u != unit);
+                // Add new unit to targets
+                addTarget(newUnit, state, underworld);
+              }
+            } else {
+              console.error("Polymorph failed to choose a new unit type.");
+            }
+          } else if (isPickup(target)) {
+            const pickup = target as IPickup;
+            // Don't polymorph purple portals
+            if (pickup.name == Pickup.PORTAL_PURPLE_NAME) continue;
+
+            let possiblePickupTypes = pickups.filter(p => isModActive(p, underworld) && p.name != pickup.name
+              && p.name != Pickup.PORTAL_PURPLE_NAME && p.name != Pickup.RECALL_POINT);
+
+            // We have to seed this to prevent multiplayer desync
+            const seed = seedrandom(`${underworld.turn_number} -${pickup.id} `);
+            const chosenPickupType = chooseOneOfSeeded(possiblePickupTypes, seed);
+            if (chosenPickupType) {
+              // Replace current pickup with new pickup
+              const newPickup = polymorphPickup(pickup, chosenPickupType, underworld, prediction);
+              if (newPickup) {
+                // Visual aid only
+                if (prediction) {
+                  // TODO - Show new pickup?
+                }
+                // Cleanup old pickup and remove it from targets
+                Pickup.removePickup(pickup, underworld, prediction);
+                state.targetedPickups = state.targetedPickups.filter(p => p != pickup);
+                // Add new pickup to targets
+                addTarget(newPickup, state, underworld);
+              }
+            } else {
+              console.error("Polymorph failed to choose a new unit type.");
+            }
+          }
+        }
+      }
+
+      return state;
+    },
+  },
+};
+
+function polymorphUnit(fromUnit: Unit.IUnit, toUnitId: string, underworld: Underworld, prediction: boolean): Unit.IUnit | undefined {
```

Comment:
We have to handle player units here. If fromUnit is of type player_controlled then we just change the image and don't do Unit.create